### PR TITLE
Fail early if getExtendedNominal is called before extension has been bound.

### PR DESCRIPTION
**Pull request by davidungar**
Trip an assertion if code in the compiler directly or indirectly attempts to `getExtendedNominal` too soon (i.e. before the extension has been bound).
Requires https://github.com/apple/swift-lldb/pull/2021

### CODE REVIEW

**Comment by davidungar** [comment]
@slavapestov @CodaFi Hasn't been tested yet, but thought you might want to get a jump on the review. Thanks.

**Comment by davidungar** [comment]
@swift-ci please smoke test os x platform

**Comment by davidungar** [comment]
@swift-ci please smoke test os x platform

**Review by CodaFi — CHANGES_REQUESTED** [review]
This infrastructure change doesn't seem quite right.  Do you want to assert `extension->alreadyBoundToNominal()`?

**Comment by davidungar** [comment]
@CodaFi How would `alreadyBoundToNominal` work in the case where the code is illegal because there is no nominal to bind the extension to?

**Comment by davidungar** [comment]
@swift-ci please smoke test os x platform

**Comment by davidungar** [comment]
@CodaFi I did change `ExtendedNominalRequest::getCachedResult` because it seems that the operation must be repeated if it fails, or else Swift.o won't build, as per your comment in the code.

**Comment by davidungar** [comment]
@swift-ci please smoke test os x platform

**Comment by davidungar** [comment]
@CodaFi Don't have time to push a fix now, but I see that the assertion must include something more as a disjunct to handle non-top-level extensions which never get attempted to be bound. The problem I see with `isInvalid() || alreadyBoundToNominal()` is the window between `bindExtensions()` and `validateDecl`. In that window, the assertion should succeed for an extension of a nonexistent nominal, but `isInvalid` won't be true yet. It's too late to mess with this now; I'll try a specific-to-not-top-level fix tomorrow (really later today).

File: include/swift/AST/Decl.h [hunk 1]
```diff
@@ -1677,7 +1677,12 @@ class ExtensionDecl final : public GenericContext, public Decl,
   TypeRepr *ExtendedTypeRepr;
 
   /// The nominal type being extended.
-  NominalTypeDecl *ExtendedNominal = nullptr;
+  ///
+  /// The bit indicates whether binding has been attempted. The pointer can be
+  /// null if either no binding was attempted or if binding could not find  the
+  /// extended nominal.
+  llvm::PointerIntPair<NominalTypeDecl *, 1, bool> ExtendedNominal = {nullptr,
```
Comment (slavapestov) [inline include/swift/AST/Decl.h]: PointerIntPair is default-initialized

**Review by slavapestov — APPROVED** [review]
This is great, thanks! I only have a few minor nitpicks, feel free to merge without more review (assuming nobody else has objections)

File: include/swift/AST/Decl.h (hunk 1 above)
Comment (davidungar) [reply include/swift/AST/Decl.h]: Cool, tx!

File: lib/AST/NameLookupRequests.cpp [hunk 2]
```diff
@@ -74,17 +74,16 @@ Optional<NominalTypeDecl *> ExtendedNominalRequest::getCachedResult() const {
   // Note: if we fail to compute any nominal declaration, it's considered
   // a cache miss. This allows us to recompute the extended nominal types
   // during extension binding.
+  // Without this recomputation, building Swift.o fails.
```
[the remark this answers is not in the document]
Comment (davidungar) [reply lib/AST/NameLookupRequests.cpp]: Edited the comment. Better?

File: lib/AST/Decl.cpp [hunk 3]
```diff
@@ -1076,9 +1076,21 @@ ExtensionDecl::takeConformanceLoaderSlow() {
...
-  return evaluateOrDefault(ctx.evaluator,
-    ExtendedNominalRequest{const_cast<ExtensionDecl *>(this)}, nullptr);
+  (void)evaluateOrDefault(
+      ctx.evaluator, ExtendedNominalRequest{const_cast<ExtensionDecl *>(this)},
+      nullptr);
+  return getExtendedNominal();
+}
+
+bool ExtensionDecl::canNeverBeBound() const {
+  return getParent()->getContextKind() != DeclContextKind::FileUnit;
```
[the remark this answers is not in the document]
Comment (davidungar) [reply lib/AST/Decl.cpp]: As discussed offline, also need to catch `struct S { extension E {} }`

File: include/swift/AST/Decl.h [hunk 4]
```diff
@@ -1744,8 +1755,18 @@ class ExtensionDecl final : public GenericContext, public Decl,
   Type getExtendedType() const;
 
   /// Retrieve the nominal type declaration that is being extended.
+  /// Will  trip an assertion if the declaration has not already been computed..
+  /// in order to fail fast when type checking work is attempted before extension binding has taken place.
```
[the remark this answers is not in the document]
Comment (davidungar) [reply include/swift/AST/Decl.h]: Thanks for the catch!

**Comment by davidungar** [comment]
@swift-ci please smoke test os x platform

**Comment by davidungar** [comment]
@swift-ci please smoke test os x platform

**Comment by davidungar** [comment]
@swift-ci please smoke test os x platform

**Comment by davidungar** [comment]
@swift-ci please clean smoke test os x platform

**Comment by davidungar** [comment]
@swift-ci please smoke test os x platform

**Comment by davidungar** [comment]
@swift-ci please smoke test os x platform

**Comment by davidungar** [comment]
@swift-ci please smoke test linux platform

**Comment by davidungar** [comment]
Please test with following pull request:
https://github.com/apple/swift-lldb/pull/2021

@swift-ci please smoke test os x platform

**Comment by davidungar** [comment]
@swift-ci please smoke test linux platform

**Comment by davidungar** [comment]
Please test with following pull request:
apple/swift-lldb#2021

@swift-ci please smoke test linux platform

**Pull request merged**